src/magic_folder/web.py

```diff
@@ -41,6 +151,36 @@ def error(request, code, message):
     request.setResponseCode(code, message)
     request.finish()
 
+
+def _is_authorized(request, get_auth_token):
+    """
+    Check the authorization carried by the given request.
+
+    :param IRequest request: The request object being considered.  If it has
+        an *Authorization* header carrying a *Bearer token* matching the token
+        returned by ``get_auth_token`` the request is considered authorized,
+        otherwise it is not.
+
+    :param (IO bytes) get_auth_token: A callable which returns the
+        authorization token value which is required for requests to be
+        authorized.
+
+    :return bool: True if and only if the given request contains the required
+        authorization materials.
+    """
+    authorization = request.requestHeaders.getRawHeaders(u"authorization")
+    if authorization is None or len(authorization) == 0:
+        return False
+    if len(authorization) > 1:
+        return False
+    auth_token = get_auth_token()
+    expected = u"Bearer {}".format(auth_token).encode("ascii")
+    return timing_safe_compare(
```

Comment:
We're depending on this being "not just binary data" on disk then I guess? (It looks like tahoe-lafs -- whose token we're currently re-using -- is currently using urlsafe-base64 encoding on-disk so that should be fine)

Reply:
Yea ... I started thinking about encoding and then thought there was no reason ... then Hypothesis told me my headers were being populated with incompatible garbage.  So, indeed, we do need the token to be "HTTP header safe".

The approach taken by the code now is to ignore this fact and just require the token read from disk to be okay.  I'd be okay with adding some safeguards here.  What would make sense?  Automatically encode the token read from disk?  Then it doesn't matter what's it disk, we can make it safe.  Clients have to agree with that encoding, of course.

Reply:
I think the *easiest* thing is to just say the token is "literally all the bytes in the file on disc" and we also ensure that's "safe" -- I did just adjust the config-PR to reflect this (so "our" on-disk token will also use urlsafe-base64 like tahoe does). This way clients inter-operating don't have to figure out how to correctly encode "whatever they found on disk" (e.g. 32 bytes) and instead just load it all and stuff it in a header.


No action for this PR, but: we should probably double-check with our cryptographers about this whole scheme. Tor for example uses the on-disk token as a key to an HMAC (of a fixed string plus a server nonce). But, that's a two-step handshake. I guess we'd have to decide if "a thing that can sniff local traffic but NOT read local files" is a threat we care about...

Reply:
Filed #202